src/Hazelcast.Net/Networking/AddressProvider.cs

```diff
@@ -68,49 +85,73 @@ public AddressProvider(NetworkingOptions networkingOptions, ILoggerFactory logge
             }
         }
 
+        /// <summary>
+        /// Whether the address provider has a map of internal addresses to public addresses.
+        /// </summary>
+        public bool HasMap { get; }
+
         /// <summary>
         /// Gets known possible addresses for a cluster.
         /// </summary>
         /// <returns>All addresses.</returns>
         public IEnumerable<NetworkAddress> GetAddresses()
         {
-            _privateToPublic ??= _createMap() ?? throw new HazelcastException("Failed to obtain an address map.");
-            return _privateToPublic.Keys;
+            _internalToPublicMap ??= _createMap() ?? throw new HazelcastException("Failed to obtain addresses.");
```

Comment:
I think that there may be another more important problem here. [Java client reloads all the addresses](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java#L559) in each call but you just use the cached one here, right? This means that `cloudScanner.Scan()` is only called once, but we expect it to be called each time the client is disconnected from cluster to re-connect to the cluster.

Reply:
`a ??= b` is equivalent to `if (a == null) { a = b; }`
`a = b ?? c` is equivalent to `a = b` if `c` is not null, else use `c` / do `c`

so : if we have a map already, do nothing. if we do not have a map, create it, and if it's created `null`, throw

Reply:
Let's check these things:
* in Java, we reload the address map if we get disconnected.
we should reset the map in .NET too when disconnected - then it will be fetched again.

* re. `?? throw` - it may not happen today but may happen in the future
but we need a way to make this coverable

Reply:
Update:

I have pushed a commit that deals with the second point (code coverage etc)

As for "reloading the map if disconnected" here what I see in Java:

There is a `RemoteAddressProvider` class that obtains addresses from Cloud Discovery (`privateToPublic = getAddresses.call()`) once and then uses them. If it fails to translate an address, it will run discovery again. But as long as it can translate addresses, regardless of whether the client disconnects, it keeps the same addreses.

.NET works the same.

That address provider is created by `ClusterDiscoveryServiceBuilder.build` which creates the `RemoteAddressProvider` once and stick with it, never changes it. `build` is invoked from `HazelcastClientInstanceImpl` `initClusterDiscoveryService` which itself is invoked from the constructor.

So... unless we create a new client, I fail to see where, should the client disconnect, the address map would be re-fetched. As far as I can see the address map would be re-fetched only if we try and fail to translate an addfress. And .NET implements this, see the `Map` method.

Reply:
In Java, `privateToPublic` is modified in two places `loadAddresses` and `translate`. .NET is missing what is being done in  `loadAddresses`, right?

Reply:
In Java, `privateToPublic` is modified in `loadAddresses` (always re-generated) and in `translate` (re-generated if mapping fails). And, `loadAddresses` is invoked by `getPossibleMemberAddresses` each time we get addresses to try to connect the client to the cluster.

In .NET, `_internalToPublicMap` is modified via `EnsureMap` in `GetAddresses` *if* it does not exist already, and in `Map` (re-generated if mapping fails). 

So... my fix broke it, and now we're not in-line with Java anymore. Or, let's not say "with Java", let's say, with some "specs" that say that (1) get a new map anytime you connect the client and cache it and (2) reuse it to map, unless map fails, then try to get a new map.

Pushing a fix now.

Reply:
Pushed + added tests to verify that :
* the map is updated each time `GetAddresses` is invoked
* the map is then cached and reused when mapping
* the map is created once and cached, when mapping, if it wasn't existing yet
* if mapping fails, and the map existed, the map is re-created (second chance)
* if mapping fails, and the map did not exist and was just created, the map is not re-created

Reply:
Are these 3 cases for when connecting to cluster? I don't understand how caching relates to the cluster connection process. Caching can relate only when we do a map.put/get kind of operations but those should also be using the addresses from the cluster member list. Is this an extra feature here? Does it exist in Java? Let's talk about it. 

Otherwise, the fix looks good.